### PR TITLE
8301341: LinkedTransferQueue does not respect timeout for poll()

### PR DESCRIPTION
**Pull request by robm-openjdk**
Backport

- 830134
- 8300663
- 8267502

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] [JDK-8301341](https://bugs.openjdk.org/browse/JDK-8301341) needs maintainer approval
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] [JDK-8300663](https://bugs.openjdk.org/browse/JDK-8300663) needs maintainer approval
- [x] [JDK-8267502](https://bugs.openjdk.org/browse/JDK-8267502) needs maintainer approval

### Issues
 * [JDK-8301341](https://bugs.openjdk.org/browse/JDK-8301341): LinkedTransferQueue does not respect timeout for poll() (**Bug** - P4 - Approved)
 * [JDK-8300663](https://bugs.openjdk.org/browse/JDK-8300663): java/util/concurrent/SynchronousQueue/Fairness.java failed with "Error: fair=true i=0 j=1" (**Bug** - P3 - Approved)
 * [JDK-8267502](https://bugs.openjdk.org/browse/JDK-8267502): JDK-8246677 caused 16x performance regression in SynchronousQueue (**Bug** - P3 - Approved)

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk21u.git pull/168/head:pull/168` \
`$ git checkout pull/168`

Update a local copy of the PR: \
`$ git checkout pull/168` \
`$ git pull https://git.openjdk.org/jdk21u.git pull/168/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 168`

View PR using the GUI difftool: \
`$ git pr show -t 168`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk21u/pull/168.diff">https://git.openjdk.org/jdk21u/pull/168.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk21u/pull/168#issuecomment-1721700296)

### CODE REVIEW

File: src/java.base/share/classes/java/util/concurrent/LinkedTransferQueue.java [hunk 1]
```diff
@@ -178,143 +172,43 @@ public class LinkedTransferQueue<E> extends AbstractQueue<E>
...
-     * pointer updates: Threads may sometimes shortcut the "next" link
-     * path from the current "head" node to be closer to the currently
-     * known first unmatched node, and similarly for tail. Again, this
-     * may be triggered with using thresholds or randomization.
+     * similarly for "tail") is an empirical matter. Larger values
+     * introduce increasing costs of cache misses and risks of long
+     * traversal chains and out-of-order updates, while smaller values
+     * increase CAS contention and overhead. Using the smallest
+     * non-zero value of one is both simple and empirically a good
+     * choice in most applicatkions.  The slack value is hard-wired: a
```
Comment (Martin-Buchholz) [inline src/java.base/share/classes/java/util/concurrent/LinkedTransferQueue.java:180]: typo: applicatkions

File: src/java.base/share/classes/java/util/concurrent/LinkedTransferQueue.java [hunk 2]
```diff
@@ -350,337 +242,472 @@ public class LinkedTransferQueue<E> extends AbstractQueue<E>
...
+     * blocking support, untimed waits use ManagedBlockers when
+     * callers are ForkJoin threads, but timed waits use plain
+     * parkNanos, under the rationale that known-to-be transient
+     * blocking doesn't require compensation. (This decision should be
+     * revisited here and elsewhere to deal with very long timeouts.)
+     *
+     * 5. Park/unpark signalling otherwise relies on a Dekker-like
+     * scheme in which the caller advertises the need to unpark by
+     * setting its waiter field, followed by a full fence and recheck
+     * before actually parking. An explicit fence in used here rather
```
Comment (Martin-Buchholz) [inline src/java.base/share/classes/java/util/concurrent/LinkedTransferQueue.java:346]: in used -> is used

**Review by Martin-Buchholz — COMMENTED** [review]
OOPS - I shouldn't have looked at this - didn't reallize it was a backport

**Comment by viktorklang-ora** [comment]
I don't see any problem backporting this. /cc @AlanBateman

**Review by AlanBateman — APPROVED** [review]

**Pull request closed without merging**